pkg/server/config.go

```diff
@@ -185,6 +177,45 @@ func NewConfig(opts *kcpserveroptions.CompletedOptions) (*Config, error) {
 
 	c.GenericConfig.RequestInfoResolver = requestinfo.NewFactory() // must be set here early to avoid a crash in the EnableMultiCluster roundtrip wrapper
 
+	// a note on the identities in a multi-shard environment
+	//
+	// case 1: read the identities from the local instance when we are on the root shard
+	// case 2: read the identities from the local instance when we are on the root shard and the caching layer was provided
+	// case 3: read the identities from the root shard when we are on a non-root shard and the caching layer and RootShardKubeconfigFile was provided
+	//
+	// case 3: RootShardKubeconfigFile was provided which indicates we are on a non-root shard
+	// thus we need to establish a connection to the root shard for identities
+	//
+	// invalid cases:
+	//
+	// case 4: we are on a non-root shard and the caching layer was provided
+	// we don't have enough info to read the identities
+	//
+	// case 5: we are on a non-root shard and RootShardKubeconfigFile was provided
+	// even though we can read the identities we won't be able to read actual resources
+	// since these are synced from the CacheKcpSharedInformerFactory first to KcpSharedInformerFactory (there is a controller)
+	var kcpIdentityRoundTripper func(rt http.RoundTripper) http.RoundTripper
+	kcpIdentityRoundTripper, c.resolveIdentities = bootstrap.NewWildcardIdentitiesWrappingRoundTripper(bootstrap.KcpRootGroupExportNames, bootstrap.KcpRootGroupResourceExportNames, c.GenericConfig.LoopbackClientConfig, nil)
+	c.identityConfig = rest.CopyConfig(c.GenericConfig.LoopbackClientConfig)
+	c.identityConfig.Wrap(kcpIdentityRoundTripper)
+	// create an empty non-functional factory so that code that uses it but doesn't need it, doesn't have to check against the nil value
+	c.CacheKcpSharedInformerFactory = kcpinformers.NewSharedInformerFactory(nil, resyncPeriod)
+	if len(c.Options.Extra.RootShardKubeconfigFile) > 0 {
+		// TODO(p0lyn0mial): use kcp-admin instead of system:admin
+		nonIdentityRootKcpShardSystemAdminConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{ExplicitPath: c.Options.Extra.RootShardKubeconfigFile}, &clientcmd.ConfigOverrides{CurrentContext: "system:admin"}).ClientConfig()
+		if err != nil {
+			return nil, fmt.Errorf("failed to load the kubeconfig from: %s, for the root shard, err: %w", c.Options.Extra.RootShardKubeconfigFile, err)
+		}
+		kcpIdentityRoundTripper, c.resolveIdentities = bootstrap.NewWildcardIdentitiesWrappingRoundTripper(bootstrap.KcpRootGroupExportNames, bootstrap.KcpRootGroupResourceExportNames, nonIdentityRootKcpShardSystemAdminConfig, c.KubeClusterClient)
+		identityRootKcpShardSystemAdminConfig := rest.CopyConfig(nonIdentityRootKcpShardSystemAdminConfig)
+		identityRootKcpShardSystemAdminConfig.Wrap(kcpIdentityRoundTripper)
+		c.RootShardKcpClusterClient, err = kcpclient.NewClusterForConfig(identityRootKcpShardSystemAdminConfig)
```

Comment:
Why do we bother to make the factory non-nil in the default case, but not the client(s)?

Reply:
because some controllers accept the factory, making it non-nil avoids constantly checking against `nil`

Reply:
Right, I meant the other case - if the implicit API here is "cache stuff is never nil", shouldn't we also make the clients use some dummy impl?